python/tank/bootstrap/manager.py

```diff
@@ -574,14 +593,109 @@ def _bootstrap_sgtk(self, engine_name, entity, progress_callback=None):
+        # Only return the Toolkit instance.
+        return self._bootstrap_toolkit_internal(entity, None)[0]
+
+    def _bootstrap_toolkit_and_cache_apps(self, engine_name, entity, progress_callback):
+        """
+        Create an :class:`~sgtk.Sgtk` instance for the given entity.
+
+        If entity is None, the method will bootstrap into the site
+        config. This method will attempt to resolve the configuration and download it
+        locally. It will also cache the application bundles.
+
+        :param engine_name: Name of the engine used to find which applications need to be cached.
+        :param entity: Shotgun entity used to resolve a project context.
+        :type entity: Dictionary with keys ``type`` and ``id``, or ``None`` for the site
+        :param progress_callback: Callback function that reports back on the toolkit bootstrap progress.
+                                  Set to ``None`` to use the default callback function.
+        :returns: Bootstrapped :class:`~sgtk.Sgtk` instance.
+        """
+
+        # First get the Toolkit instance
+        tk, status = self._bootstrap_toolkit_internal(entity, progress_callback)
+
+        # Now cache the apps
+        if status != Configuration.LOCAL_CFG_UP_TO_DATE:
+            self._cache_apps(tk, tk.pipeline_configuration, engine_name, entity, progress_callback)
+
+        return tk
+
+    def _bootstrap_toolkit_internal(self, entity, progress_callback):
+        """
+        Create an :class:`~sgtk.Sgtk` instance for the given entity.
+
+        If entity is None, the method will bootstrap into the site
+        config. This method will attempt to resolve the configuration and download it
+        locally. Note that it will not cache the application bundles.
+
+        Please note that the API version of the :class:`~sgtk.Sgtk` instance may not be the same as the
+        API version that was used during the bootstrap.
+
+        :param entity: Shotgun entity used to resolve a project context.
+        :type entity: Dictionary with keys ``type`` and ``id``, or ``None`` for the site
+        :param progress_callback: Callback function that reports back on the toolkit bootstrap progress.
+                                  Set to ``None`` to use the default callback function.
+        :returns: Bootstrapped :class:`~sgtk.Sgtk` instance and the configuration's status.
+        """
+
+        #
+        if progress_callback is None:
+            progress_callback = self.progress_callback
+
+        config, status = self._get_configuration(entity, progress_callback)
+
         # we can now boot up this config.
         self._report_progress(progress_callback, 0.3, "Starting up Toolkit...")
         tk = config.get_tk_instance(self._sg_user)
 
-        if status != Configuration.LOCAL_CFG_UP_TO_DATE:
-            self._cache_apps(tk, engine_name, entity, progress_callback)
+        return tk, status
 
-        return tk
+    def update_and_cache_configuration(self, project):
```

Comment:
`prepare_engine` can't take an entity in because it won't be able to run the pick environment hook. Although in theory it could resolve the project from the entity, but I decided against it. Shall I add that in?

As for making the functionality symmetric, I agree.